File: scripts/ids_to_yaml.py
#!/usr/bin/env python3

import argparse
import yaml


def main():
    parser = argparse.ArgumentParser(description="Convert a list of items into a structured YAML format.")
    parser.add_argument("input_file", help="Path to the IDS list.")
    parser.add_argument("doc_value", help="IDS filename to display in tooltips")
    parser.add_argument("output_file", help="Path to the output YAML file.")

    args = parser.parse_args()

    with open(args.input_file, "r", encoding="utf-8") as file:
        lines = file.readlines()

    data = []
    for line in lines:
        line = line.strip()
        if line:  # Skip empty lines
            parts = line.split()
            if len(parts) == 2:
                value, name = parts
                data.append({"name": name, "detail": value, "doc": args.doc_value})

    with open(args.output_file, "w", encoding="utf-8") as file:
        yaml.dump(data, file, default_flow_style=False, sort_keys=False)

    print(f"Output written to {args.output_file}")


if __name__ == "__main__":
    main()

File: scripts/utils/src/ids-to-yaml.ts
/**
 * Converts IDS (Infinity Engine Data Structure) list files to YAML
 * completion data format.
 *
 * Replaces the former ids_to_yaml.py script.
 *
 * Usage:
 *   pnpm exec tsx scripts/utils/src/ids-to-yaml.ts <input.ids> <doc_value> <output.yml>
 */

import fs from "node:fs";
import YAML from "yaml";

interface IdsEntry {
    readonly name: string;
    readonly detail: string;
    readonly doc: string;
}

/**
 * Parses an IDS file into structured entries.
 * Each non-empty line is expected to have format: `<value> <name>`.
 * Lines that don't match this format are skipped.
 */
export function parseIdsFile(content: string, docValue: string): readonly IdsEntry[] {
    const entries: IdsEntry[] = [];
    for (const rawLine of content.split("\n")) {
        const line = rawLine.trim();
        if (line === "") {
            continue;
        }
        const parts = line.split(/\s+/);
        if (parts.length === 2) {
            entries.push({
                name: parts[1]!,
                detail: parts[0]!,
                doc: docValue,
            });
        }
    }
    return entries;
}

// -- CLI entry point (tested via subprocess in ids-to-yaml-cli.test.ts) --

/* v8 ignore start -- CLI wrapper tested via execSync integration tests */
function main(): void {
    const inputFile = process.argv[2];
    const docValue = process.argv[3];
    const outputFile = process.argv[4];

    if (inputFile === undefined || docValue === undefined || outputFile === undefined) {
        console.error("Usage: ids-to-yaml <input.ids> <doc_value> <output.yml>");
        process.exit(1);
    }

    const content = fs.readFileSync(inputFile, "utf8");
    const entries = parseIdsFile(content, docValue);
    const yamlStr = YAML.stringify(entries);
    fs.writeFileSync(outputFile, yamlStr, "utf8");

    console.log(`Output written to ${outputFile}`);
}

const isDirectRun = process.argv[1]?.endsWith("ids-to-yaml.ts");
if (isDirectRun) {
    main();
}
/* v8 ignore stop */
